utils/node_local_test.py

```diff
@@ -42,20 +42,149 @@ class NLT_Conf():
     """Helper class for configuration"""
     def __init__(self, bc):
         self.bc = bc
-        self.output_fd = None
         self.agent_dir = None
+        self.wf = None
 
-    def set_output_file(self, filename):
-        """Set the name of a output file for error logging
-
-        This is used to save the lines or src that report issues
-        for use with the Jenkins warnings-ng plugin
-        """
-        self.output_fd = open(filename, 'w')
+    def set_wf(self, wf):
+        """Set the WarningsFactory object"""
+        self.wf = wf
 
     def __getitem__(self, key):
         return self.bc[key]
 
+class WarningsFactory():
+    """Class to parse warnings, and save to JSON output file
+
+    Take a list of failures, and output the data in a way that is best
+    displayed according to
+    https://github.com/jenkinsci/warnings-ng-plugin/blob/master/doc/Documentation.md
+    """
+
+    # Error levels are LOW, NORMAL, HIGH, ERROR
+    FLAKY_FUNCTIONS = ('daos_lru_cache_destroy', 'vos_tls_fini', 'rdb_timerd')
+
+    def __init__(self, filename):
+        self._fd = open(filename, 'w')
+        self.issues = []
+        self.pending = []
+        self._running = True
+        # Save the filename of the object, as __file__ does not
+        # work in __del__
+        self._file = __file__.lstrip('./')
+        self._flush()
+
+    def __del__(self):
+        """Ensure the file is flushed on exit, but if it hasn't already
+        been closed then mark an error"""
+        if not self._fd:
+            return
+
+        entry = {}
+        entry['fileName'] = os.path.basename(self._file)
+        entry['directory'] = os.path.dirname(self._file)
+        entry['lineStart'] = sys._getframe().f_lineno
```

Comment:
(pylint-protected-access) Access to a protected member _getframe of a client class